pkg/server/server.go

```diff
@@ -19,57 +22,79 @@ type Service interface {
 	Run(ctx context.Context, server *Server) error
 }
 
+type handle struct {
+	grpc *grpc.Server
+	lis  net.Listener
+}
+
+func (h handle) Serve() error {
+	err := h.grpc.Serve(h.lis)
+	if err == grpc.ErrServerStopped {
+		return nil
+	}
+	return Error.Wrap(err)
+}
+
+func (h handle) Close() error {
+	h.grpc.GracefulStop()
+	// GracefulStop closes the listener
+	return nil
+}
+
 // Server represents a bundle of services defined by a specific ID.
 // Examples of servers are the satellite, the storagenode, and the uplink.
 type Server struct {
-	lis      net.Listener
-	grpc     *grpc.Server
+	public   handle
+	private  handle
 	next     []Service
 	identity *identity.FullIdentity
 }
 
-// NewServer creates a Server out of an Identity, a net.Listener,
-// a UnaryServerInterceptor, and a set of services.
-func NewServer(opts *Options, lis net.Listener,
-	interceptor grpc.UnaryServerInterceptor, services ...Service) (
-	*Server, error) {
-	grpcOpts, err := opts.grpcOpts()
-	if err != nil {
-		return nil, err
-	}
-
-	unaryInterceptor := unaryInterceptor
-	if interceptor != nil {
-		unaryInterceptor = combineInterceptors(unaryInterceptor, interceptor)
-	}
-
+// NewServer creates a Server out of an Identity, public and private
+// gRPC handles and listeners, and a set of services. Closing a server will
+// stop the grpc servers and close the listeners.
+// A public handler is expected to be exposed to the world, whereas the private
+// handler is for inspectors and debug tools only.
+func NewServer(
+	identity *identity.FullIdentity,
+	publicSrv *grpc.Server, publicLis net.Listener,
+	privateSrv *grpc.Server, privateLis net.Listener,
+	services ...Service) *Server {
```

Comment:
remove line breaks
also can this take instead a `identity *identity.FullIdentity, public, private Handle, services ...Service`